text/0000-debugger-visualizer.md

```diff
@@ -0,0 +1,643 @@
+set for each `.natvis` file which will embed the Natvis visualizations into the PDB.
+
+To provide Natvis files, developers create a file with the `.natvis` file
+extension.
+
+See the below example for how rustc would embed these Natvis files in the
+debuginfo for a binary.
+
+As an example, consider a crate `foo` with this directory structure:
+
+```text
+/Cargo.toml
+/Foo.natvis (Note: the .natvis file does not have to match the name of the crate.)
+  +-- src
+      +-- main.rs
+```
+
+Where `main.rs` contains:
+
+```rust
+#[debugger_visualizer(natvis_file = "../Foo.natvis")]
+
+/// A rectangle in first quadrant
+struct FancyRect {
+    pub x: f32,
+    pub y: f32,
+    pub dx: f32,
+    pub dy: f32,
+}
+
+fn main() {
+    let mut fancy_rect = FancyRect::new(10.0, 10.0, 5.0, 5.0);
+    println!("FancyRect: {:?}", fancy_rect);
+}
+```
+
+and `Foo.natvis` contains:
+
+```xml
+<?xml version="1.0" encoding="utf-8"?>
+<AutoVisualizer xmlns="http://schemas.microsoft.com/vstudio/debugger/natvis/2010">
+    <Type Name="foo::FancyRect">
+      <DisplayString>({x},{y}) + ({dx}, {dy})</DisplayString>
+      <Expand>
+        <Synthetic Name="LowerLeft">
+          <DisplayString>({x}, {y})</DisplayString>
+        </Synthetic>
+        <Synthetic Name="UpperLeft">
+          <DisplayString>({x}, {y + dy})</DisplayString>
+        </Synthetic>
+        <Synthetic Name="UpperRight">
+          <DisplayString>({x + dx}, {y + dy})</DisplayString>
+        </Synthetic>
+        <Synthetic Name="LowerRight">
+          <DisplayString>({x + dx}, {y})</DisplayString>
+        </Synthetic>
+      </Expand>
+    </Type>
+</AutoVisualizer>
+```
+
+When viewed under WinDbg, the `fancy_rect` variable would be shown as follows:
+
+```text
+> Variables:
+  > fancy_rect: (10, 10) + (5, 5)
+    > LowerLeft: (10, 10)
+    > UpperLeft: (10, 15)
+    > UpperRight: (15, 15)
+    > LowerRight: (15, 10)
+```
+
+## Supporting Pretty Printers
+
+This section describes how GDB's pretty printers are supported in Rust.
+
+To use a pretty printer, developers write python scripts that describe how a type
+should be displayed when loaded up in GDB/LLDB. (See: https://sourceware.org/gdb/onlinedocs/gdb/Pretty-Printing.html#Pretty-Printing)
+The pretty printers provide patterns, which match type names, and for matching
+types, descibe how to display those types. (For writing a pretty printer, see: https://sourceware.org/gdb/onlinedocs/gdb/Writing-a-Pretty_002dPrinter.html#Writing-a-Pretty_002dPrinter).
+
+Rust developers can add one or more pretty printers to their crate. This is done
+in the Rust compiler via `.py` python scripts. Through the use of a new Rust attribute,
+`#[debugger_visualizer]`, the compiler will encode the contents of the `.py` file in
+the crate metadata if the target is an `rlib`. If the target is an executable, the
+`.debug_gdb_scripts` section will include a reference to the pretty printer specified.
```

Comment:
> But I think it might be generally more robust to use the "embed script contents" version.

I totally agree. Depending on files in search paths is not always the most reliable method. The contents of the script is already embedded in the crate metadata, it should be very simple to directly embed that instead of the file path. I will update the RFC to reflect this change.